services/src/main/java/org/keycloak/authentication/AuthenticatorUtil.java

```diff
@@ -94,7 +94,12 @@ public static Set<String> getAuthCallbacksFactoryIds(AuthenticationSessionModel
         final String callbacksFactories = authSession.getAuthNote(CALLBACKS_FACTORY_IDS_NOTE);
 
         if (StringUtil.isNotBlank(callbacksFactories)) {
-            return Sets.newHashSet(callbacksFactories.split(Constants.CFG_DELIMITER));
+            String[] split = callbacksFactories.split(Constants.CFG_DELIMITER);
+            Set<String> set = new HashSet<>(split.length);
+            for (String s : split) {
+                set.add(s);
+            }
+            return Collections.unmodifiableSet(set);
```

Comment:
`Sets.newHashSet` does not return an immutable set. Not a blocker if it makes sense here.

Reply:
True, but did it since there isn't much point in modifying the returned set as it won't make any difference on the underlying note.